src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1015,3 +1057,958 @@ def upgrade_agents(cmd, client, resource_group_name, cluster_name, kube_config=N
+    else:
+        auth_method = 'AAD'
+        telemetry.add_extension_event('connectedk8s', {'Context.Default.AzureCLI.IsAADEnabled': True})
+
+    # Fetching hybrid connection details from Userrp
+    try:
+        response = client.list_cluster_user_credentials(resource_group_name, cluster_name, auth_method, True)
+    except Exception as e:
+        utils.arm_exception_handler(e, consts.Get_Credentials_Failed_Fault_Type, 'Unable to list cluster user credentials')
+        if flag == 0:
+            raise CLIError("Failed to get credentials." + str(e))    
+        else:
+            close_subprocess_and_raise_cli_error(clientproxy_process, "Failed to get credentials." + str(e))
+    
+    # Starting the client proxy process, if this is the first time that this function is invoked
+    if flag == 0:
+        try:
+            if debug_mode:
+                clientproxy_process = Popen(args)
+            else:
+                clientproxy_process = Popen(args, stdout=DEVNULL, stderr=DEVNULL)
+            print(f'Proxy is listening on port {api_server_port}')
+
+        except Exception as e:
+            telemetry.set_exception(exception=e, fault_type=consts.Run_Clientproxy_Fault_Type,
+                                    summary='Unable to run client proxy executable')
+            raise CLIError("Failed to start proxy process." + str(e))
+
+        if user_type == 'user':
+            identity_data = {}
+            identity_data['refreshToken'] = creds
+            identity_uri = f'https://localhost:{api_server_port}/identity/rt'
+
+            # Needed to prevent skip tls warning from printing to the console
+            original_stderr = sys.stderr
+            f = open(os.devnull, 'w')
+            sys.stderr = f
+
+            make_api_call_with_retries(identity_uri, identity_data, False, consts.Post_RefreshToken_Fault_Type,
+                                       'Unable to post refresh token details to clientproxy',
+                                       "Failed to pass refresh token details to proxy.", clientproxy_process)
+            sys.stderr = original_stderr
+
+    data = prepare_clientproxy_data(response)
+    expiry = data['hybridConnectionConfig']['expirationTime']
+
+    if token is not None:
+        data['kubeconfigs'][0]['value'] = insert_token_in_kubeconfig(data, token)
+
+    uri = f'http://localhost:{client_proxy_port}/subscriptions/{subscription_id}/resourceGroups/{resource_group_name}/providers/Microsoft.Kubernetes/connectedClusters/{cluster_name}/register?api-version=2020-10-01'
+
+    # Posting hybrid connection details to proxy in order to get kubeconfig
+    response = make_api_call_with_retries(uri, data, False, consts.Post_Hybridconn_Fault_Type,
+                                          'Unable to post hybrid connection details to clientproxy',
+                                          "Failed to pass hybrid connection details to proxy.", clientproxy_process)
+
+    if flag == 0:
```

Comment:
Let's verify by running long running and close this comment if only merging kubeconfig is required only once or not.

Reply:
long running has worked for more than 9 hrs. So merging kubeconfig every time is not needed.